File: abstract-factory/src/main/java/com/github/ftfetter/oopdesignpatterns/abstractfactory/Application.java
package com.github.ftfetter.oopdesignpatterns.abstractfactory;

import com.github.ftfetter.oopdesignpatterns.abstractfactory.banks.BankFactory;
import com.github.ftfetter.oopdesignpatterns.abstractfactory.banks.BlueBankFactory;
import com.github.ftfetter.oopdesignpatterns.abstractfactory.banks.YellowBankFactory;
import com.github.ftfetter.oopdesignpatterns.abstractfactory.type.BankType;

public class Application {

    public static void main(String[] args) throws Exception {
        getBank(BankType.YELLOW_BANK)
                .createCardAccount()
                .whoAmI();

        getBank(BankType.BLUE_BANK)
                .createCurrentAccount()
                .whoAmI();
    }

    private static BankFactory getBank(BankType bankType) throws Exception {
        switch (bankType) {
            case YELLOW_BANK: return new YellowBankFactory();
            case BLUE_BANK: return new BlueBankFactory();
            default: throw new Exception("Bank not found!");
        }
    }
}

File: abstract-factory/src/main/kotlin/com/github/ftfetter/oopdesignpatterns/abstractfactory/Application.kt
package com.github.ftfetter.oopdesignpatterns.abstractfactory

import com.github.ftfetter.oopdesignpatterns.abstractfactory.factory.MageFactory
import com.github.ftfetter.oopdesignpatterns.abstractfactory.factory.RangerFactory
import com.github.ftfetter.oopdesignpatterns.abstractfactory.factory.WarriorFactory

fun main() {
    val mage = ClassClient(MageFactory())
    mage.printSheet()

    val ranger = ClassClient(RangerFactory())
    ranger.printSheet()

    val warrior = ClassClient(WarriorFactory())
    warrior.printSheet()
}